core/amber/src/main/python/core/runnables/data_processor.py

```diff
@@ -29,40 +30,53 @@ def run(self) -> None:
         self._running.set()
         self._switch_context()
         while self._running.is_set():
-            self.process_tuple()
+            marker = self._context.marker_processing_manager.get_input_marker()
+            tuple_ = self._context.tuple_processing_manager.current_input_tuple
+            if marker is not None:
+                self.process_marker(marker)
+            elif tuple_ is not None:
+                self.process_tuple()
+            else:
+                raise RuntimeError("No marker or tuple to process.")
+            self._switch_context()
+
+    def process_marker(self, marker: Marker) -> None:
+        try:
+            executor = self._context.executor_manager.executor
+            port = self._context.tuple_processing_manager.get_input_port()
+            with replace_print(
+                self._context.worker_id,
+                self._context.console_message_manager.print_buf,
+            ):
+                if isinstance(marker, StartOfUpstream):
+                    self._set_output_state(executor.produce_state_on_start(port))
+                elif isinstance(marker, State):
+                    self._set_output_state(executor.process_state(marker, port))
+                elif isinstance(marker, EndOfUpstream):
+                    self._set_output_state(executor.produce_state_on_finish(port))
+                    self._set_output_tuple(executor.on_finish(port))
```

Comment:
I thought `on_finish` was removed? also, here is setting output tuple in process state?

Reply:
We never try to remove the `on_finish` callback function; it's just moved out of process_tuple. We set an output tuple when processing the EndOfUpstream(or EndOfInputPort) marker.